enhancements/workload-partitioning/rationalizing-configuration.md

```diff
@@ -0,0 +1,478 @@
+
+## Summary
+
+The initial iteration of workload partitioning focused on a short path
+to a minimum viable implementation. This enhancement describes the
+loose ends for preparing the feature for GA at a high level, and
+explains the set of other design documents that need to be written
+separately during the next iteration.
+
+## Motivation
+
+The management workload partitioning feature introduced in OCP 4.8
+enables pod workloads to be pinned to the kubelet reserved cpuset. The
+feature is enabled at install time and requires the reserved set of
+CPUs as input. That configuration is currently passed in via a
+manually created manifest containing a machine config resource. The
+set of CPUs used for management workload partitioning has to align
+with the reserved set configured by the performance-addon-operator
+(PAO) during day 2. For OCP 4.8, the onus is on the user to ensure
+these align, there is no interlock.  The current implementation is
+prone to user error because it has two entities configuring and
+managing the same data.
+
+The 4.8 implementation is also limited to single-node deployments. We
+anticipate users of highly-available clusters also wanting to minimize
+the overhead of the management components to take full advantage of
+compute capacity for their own workloads.
+
+This design document describes phase 2 of the implementation plan for
+workload partitioning to improve the user experience when configuring
+the feature and to extend support to multi-node clusters.
+
+### Goals
+
+1. Describe, at a high level, the remaining work to improve the user
+   experience and expand support to multi-node deployments for
+   architectural review and planning.
+2. Document the division of responsibility between workload management
+   in components in the OpenShift release payload and PAO.
+3. List the other enhancements that need to be written, and their
+   scopes.
+4. Minimize the complexity of installing a "normal" cluster while
+   still supporting this very special case.
+
+### Non-Goals
+
+1. Resolve all of the details for each of the next steps. We're not
+   going to talk a lot about names, API fields, or implementation
+   details here.
+2. Describe an API for (re)configuring workload partitioning in a
+   running cluster. We are going to continue to require the user to
+   enable partitioning as part of deploying the cluster for another
+   implementation phase.
+
+## Proposal
+
+### API
+
+The initial implementation is limited to single-node deployments. To
+support multi-node clusters, we need to solve a couple of concurrency
+problems. The first is providing a way for the admission hook
+responsible for mutating pod definitions to reliably know when the
+feature should be enabled. The admission hook cannot store state on
+its own, so we need a new API to indicate when the feature is
+enabled.
+
+We eventually want to extend the partitioning feature to support
```

Comment:
> @deads2k had objections to status only fields in past when we handled aws resource tags.
> 
> [openshift/api#905](https://github.com/openshift/api/pull/905)

Yes.  Status-only fields need to be able to describe

1. how they will eventually allow human change if it is needed
2. how they react to a cluster-admin setting the field to a different value

I think this API can probably describe it, but it's important to consider it here.  For instance, I think the performance add-on operator could probably re-stomp status if it were changed (needs confirmation).  To allow a human to change it, the  performance add-on operator would need to learn that it could be externally controlled.

Clusters without the PAO won't be able to express "turn it on" without a spec field.  Would a spec field update be prevented under some conditions?

Reply:
PAO cannot own this object or guarantee any reconcile loop. It is an optional operator installed by OLM.

PAO will merely know how to prepare it using render mode just like it does for MC, KubeletConfig and others to make it easy for admins. But any admin can do the same manually even without PAO.

Reply:
We talked about this offline and agreed that we will have spec & status fields but that we will block changes using an admission plugin. See 160aa5d53d648f4d043c475a44a94a17db3d1ae8